[mir-opt] asking `?`s in a more optimized fashion

Centril opened the pull request:
This PR works towards https://github.com/rust-lang/rust/issues/66234 by providing two optimization passes meant to run in sequence:

- `SimplifyArmIdentity` which transforms something like:
  ```rust
  _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
  ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
  discriminant(_LOCAL_0) = VAR_IDX;
  ```
 
  into:
 
  ```rust
  _LOCAL_0 = move _LOCAL_1
  ```

- `SimplifyBranchSame` which transforms `SwitchInt`s to identical basic blocks into a `goto` to the first reachable target.

Together, these are meant to simplify the following into just `res`:
```rust
match res {
    Ok(x) => Ok(x),
    Err(x) => Err(x),
}
```

It should be noted however that the desugaring of `?` includes a function call and so the first pass in this PR relies on inlining to substitute that function call for identity on `x`. Inlining requires `mir-opt-level=2` so this might not have any effect in perf-bot but let's find out.

r? @oli-obk -- This is WIP, but I'd appreciate feedback. :)

Centril commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 91e5a0e88e56d021a75019a6381401746f02db54 with merge fb1a35ea56157ef7f76c1451f72cda0b3be86a38...
<!-- homu: {"type":"TryBuildStarted","head_sha":"91e5a0e88e56d021a75019a6381401746f02db54","merge_sha":"fb1a35ea56157ef7f76c1451f72cda0b3be86a38"} -->

Centril commented:
@bors retry

Centril commented:
@bors retry try

bors commented:
:hourglass: Trying commit 657c945c02f263f5f449acce9201f07f9f5576f1 with merge 0113342691c2ecbb2b7fd1d083fd0631f136b1dd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"657c945c02f263f5f449acce9201f07f9f5576f1","merge_sha":"0113342691c2ecbb2b7fd1d083fd0631f136b1dd"} -->

Centril commented on src/test/mir-opt/simplify_try.rs:
```diff
@@ -0,0 +1,193 @@
+//     scope 6 {
+//     }
+//     bb0: {
+//         _2 = discriminant(_1);
```
This is a tad unfortunate that `_2` isn't optimized out.

oli-obk replied:
I thought we had some sort of DCE. If not, open an issue so we create it

wesleywiser replied:
I think this will be easy to fix. I'll open a PR tonight with that.

oli-obk replied:
There are some ppl at rustfest working on it

Centril replied:
Does that refer to https://github.com/rust-lang/rust/pull/66329 ? -- That won't help in this case since this is an unused, but reachable, local.

oli-obk replied:
No, different optimization, not sure on what repo it's being developed

Centril commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,276 @@
+            }
+        }
+
+        // We may have dead blocks now, so remvoe those.
```
```suggestion
        // We may have dead blocks now, so remove those.
```

oli-obk replied:
Not necessarily. This should only be run if anything was changed

Centril replied:
Good point; will fix.

Centril commented on src/librustc_mir/transform/mod.rs:
```diff
@@ -275,6 +276,9 @@ fn run_optimization_passes<'tcx>(
         &copy_prop::CopyPropagation,
         &simplify_branches::SimplifyBranches::new("after-copy-prop"),
         &remove_noop_landing_pads::RemoveNoopLandingPads,
+        &simplify::SimplifyCfg::new("after-remove-noo-landing-pads"),
```
```suggestion
        &simplify::SimplifyCfg::new("after-remove-noop-landing-pads"),
```

Centril commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,276 @@
+
+/*
+
+KEEPSAKE: REMOVE IF NOT NECESSARY!
```
If we are OK with the `PartialEq`s then this is garbage that can be removed.

rust-highfive commented:
The job `dist-x86_64-linux-alt` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13218&view=logs&jobId=396dc680-0b3a-5910-2395-0c692e01f85b) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13218/logs/56), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13218/logs/56)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-11T04:30:25.1341610Z [RUSTC-TIMING] rustc_traits test:false 65.160
2019-11-11T04:30:25.1342123Z    Compiling rustc_passes v0.0.0 (/checkout/src/librustc_passes)
2019-11-11T04:31:08.6176939Z [RUSTC-TIMING] rustc_passes test:false 43.476
2019-11-11T04:31:08.6231708Z    Compiling rustc_mir v0.0.0 (/checkout/src/librustc_mir)
2019-11-11T04:31:23.8394662Z error[E0382]: use of moved value: `vf_s0`
2019-11-11T04:31:23.8395143Z   --> src/librustc_mir/transform/simplify_try.rs:54:35
2019-11-11T04:31:23.8395434Z    |
2019-11-11T04:31:23.8396607Z 45 |             let (local_tmp_s0, local_1, vf_s0) = match match_get_variant_field(s0) {
2019-11-11T04:31:23.8397180Z    |                                         ----- move occurs because `vf_s0` has type `transform::simplify_try::VarField<'_>`, which does not implement the `Copy` trait
2019-11-11T04:31:23.8397656Z ...
2019-11-11T04:31:23.8398144Z 53 |             if (local_tmp_s0, vf_s0) != (local_tmp_s1, vf_s1)
2019-11-11T04:31:23.8398573Z    |                               ----- value moved here
2019-11-11T04:31:23.8399031Z 54 |                 || Some((local_0, vf_s0.var_idx)) != match_set_discr(s2)
2019-11-11T04:31:23.8399508Z    |                                   ^^^^^^^^^^^^^ value used here after move
2019-11-11T04:31:27.2401323Z error: aborting due to previous error
2019-11-11T04:31:27.2405768Z 
2019-11-11T04:31:27.2416558Z For more information about this error, try `rustc --explain E0382`.
2019-11-11T04:31:27.3742025Z [RUSTC-TIMING] rustc_mir test:false 18.745
---
2019-11-11T04:32:48.4672990Z   local time: Mon Nov 11 04:32:48 UTC 2019
2019-11-11T04:32:48.7846213Z   network time: Mon, 11 Nov 2019 04:32:48 GMT
2019-11-11T04:32:48.7847517Z == end clock drift check ==
2019-11-11T04:32:51.6983858Z 
2019-11-11T04:32:51.7095887Z ##[error]Bash exited with code '1'.
2019-11-11T04:32:51.7135441Z ##[section]Starting: Checkout
2019-11-11T04:32:51.7137620Z ==============================================================================
2019-11-11T04:32:51.7137736Z Task         : Get sources
2019-11-11T04:32:51.7137852Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13219&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13219/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13219/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-11T04:05:18.4634736Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-11T04:05:18.4824983Z ##[command]git config gc.auto 0
2019-11-11T04:05:18.4904942Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-11T04:05:18.4964558Z ##[command]git config --get-all http.proxy
2019-11-11T04:05:18.5108718Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66282/merge:refs/remotes/pull/66282/merge
---
2019-11-11T05:04:00.7890563Z .................................................................................................... 1400/9227
2019-11-11T05:04:07.1747955Z .................................................................................................... 1500/9227
2019-11-11T05:04:13.2698505Z .................................................................................................... 1600/9227
2019-11-11T05:04:22.5020119Z .................................................................................................... 1700/9227
2019-11-11T05:04:30.8971540Z ..i................................................................................................. 1800/9227
2019-11-11T05:04:37.7082616Z ......................................................................................iiiii......... 1900/9227
2019-11-11T05:04:58.9044766Z .................................................................................................... 2100/9227
2019-11-11T05:05:01.3583095Z .................................................................................................... 2200/9227
2019-11-11T05:05:03.9124670Z .................................................................................................... 2300/9227
2019-11-11T05:05:13.4960353Z .................................................................................................... 2400/9227
---
2019-11-11T05:08:05.7606566Z ..................................................................................i...............i. 4700/9227
2019-11-11T05:08:12.8329815Z .................................................................................................... 4800/9227
2019-11-11T05:08:21.9631265Z .................................................................................................... 4900/9227
2019-11-11T05:08:27.2114287Z .................................................................................................... 5000/9227
2019-11-11T05:08:38.4190631Z ....................................................................................ii.ii........... 5100/9227
2019-11-11T05:08:42.2901894Z i................................................................................................... 5200/9227
2019-11-11T05:08:57.0755168Z .................................................................................................... 5400/9227
2019-11-11T05:09:04.0494353Z ..................................................................i................................. 5500/9227
2019-11-11T05:09:12.1834079Z .................................................................................................... 5600/9227
2019-11-11T05:09:20.3521104Z .................................................................................................... 5700/9227
2019-11-11T05:09:20.3521104Z .................................................................................................... 5700/9227
2019-11-11T05:09:28.9561883Z ...................................................ii...i..ii...........i........................... 5800/9227
2019-11-11T05:09:51.6819754Z .................................................................................................... 6000/9227
2019-11-11T05:10:00.1589920Z .................................................................................................... 6100/9227
2019-11-11T05:10:00.1589920Z .................................................................................................... 6100/9227
2019-11-11T05:10:10.4098804Z ......................................................................i..ii......................... 6200/9227
2019-11-11T05:10:40.5355090Z .................................................................................................... 6400/9227
2019-11-11T05:10:42.7380818Z ......................................i............................................................. 6500/9227
2019-11-11T05:10:45.0201568Z .................................................................................................... 6600/9227
2019-11-11T05:10:47.4145733Z .......................i............................................................................ 6700/9227
---
2019-11-11T05:15:58.7979422Z  finished in 5.869
2019-11-11T05:15:58.8163757Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-11T05:15:59.0039343Z 
2019-11-11T05:15:59.0039626Z running 156 tests
2019-11-11T05:16:01.9371088Z iiii....iii......iii..iiii...i.............................i..i..................iF...i...........ii 100/156
2019-11-11T05:16:04.6807167Z .i.i..iiii..............i.........iii.i.........ii......
2019-11-11T05:16:04.6817243Z 
2019-11-11T05:16:04.6818448Z ---- [codegen] codegen/match.rs stdout ----
2019-11-11T05:16:04.6818854Z 
2019-11-11T05:16:04.6818854Z 
2019-11-11T05:16:04.6819343Z error: verification with 'FileCheck' failed
2019-11-11T05:16:04.6819577Z status: exit code: 1
2019-11-11T05:16:04.6821283Z command: "/usr/lib/llvm-6.0/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll" "/checkout/src/test/codegen/match.rs"
2019-11-11T05:16:04.6822216Z ------------------------------------------
2019-11-11T05:16:04.6822793Z 
2019-11-11T05:16:04.6823294Z ------------------------------------------
2019-11-11T05:16:04.6823537Z stderr:
2019-11-11T05:16:04.6823537Z stderr:
2019-11-11T05:16:04.6823918Z ------------------------------------------
2019-11-11T05:16:04.6824132Z /checkout/src/test/codegen/match.rs:13:11: error: expected string not found in input
2019-11-11T05:16:04.6824813Z // CHECK: switch{{.*}}, label %[[OTHERWISE:[a-zA-Z0-9_]+]] [
2019-11-11T05:16:04.6825138Z           ^
2019-11-11T05:16:04.6825661Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll:7:30: note: scanning from here
2019-11-11T05:16:04.6826006Z define void @exhaustive_match(i1 zeroext) unnamed_addr #0 {
2019-11-11T05:16:04.6826482Z                              ^
2019-11-11T05:16:04.6827032Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll:11:13: note: possible intended match here
2019-11-11T05:16:04.6827254Z  store i8 %1, i8* %e, align 1
2019-11-11T05:16:04.6827536Z 
2019-11-11T05:16:04.6828032Z ------------------------------------------
2019-11-11T05:16:04.6828248Z 
2019-11-11T05:16:04.6828372Z 
---
2019-11-11T05:16:04.6830317Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-11T05:16:04.6830689Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-11T05:16:04.6830916Z 
2019-11-11T05:16:04.6831028Z 
2019-11-11T05:16:04.6832938Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-11T05:16:04.6833445Z 
2019-11-11T05:16:04.6833585Z 
2019-11-11T05:16:04.6833727Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-11T05:16:04.6833869Z Build completed unsuccessfully in 1:04:12
2019-11-11T05:16:04.6833869Z Build completed unsuccessfully in 1:04:12
2019-11-11T05:16:04.6834029Z == clock drift check ==
2019-11-11T05:16:04.6834165Z   local time: Mon Nov 11 05:16:03 UTC 2019
2019-11-11T05:16:04.6834319Z   network time: Mon, 11 Nov 2019 05:16:03 GMT
2019-11-11T05:16:04.6834455Z == end clock drift check ==
2019-11-11T05:16:08.2521115Z 
2019-11-11T05:16:08.2620458Z ##[error]Bash exited with code '1'.
2019-11-11T05:16:08.2658395Z ##[section]Starting: Checkout
2019-11-11T05:16:08.2660347Z ==============================================================================
2019-11-11T05:16:08.2660411Z Task         : Get sources
2019-11-11T05:16:08.2660452Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13220)
Build commit: 0113342691c2ecbb2b7fd1d083fd0631f136b1dd (`0113342691c2ecbb2b7fd1d083fd0631f136b1dd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13220"},"merge_sha":"0113342691c2ecbb2b7fd1d083fd0631f136b1dd"} -->

rust-timer commented:
Queued 0113342691c2ecbb2b7fd1d083fd0631f136b1dd with parent e2fa9527d44522b900c739cc82faa734f877bcda, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e2fa9527d44522b900c739cc82faa734f877bcda&end=0113342691c2ecbb2b7fd1d083fd0631f136b1dd).

oli-obk commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,276 @@
+            // All successor basic blocks should have the exact same form.
+            let all_successors_equivalent = iter_bbs_reachable
+                .map(|(_, bb)| bb)
+                .tuple_windows()
```
If this is the only use case for itertools that seems a bit overkill to include

Centril replied:
It's already included in the lockfile and besides, if we never use it we'll just end up with boilerplate instead.

rust-timer commented:
Finished benchmarking try commit 0113342691c2ecbb2b7fd1d083fd0631f136b1dd, [comparison URL](https://perf.rust-lang.org/compare.html?start=e2fa9527d44522b900c739cc82faa734f877bcda&end=0113342691c2ecbb2b7fd1d083fd0631f136b1dd).

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13351&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13351/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13351/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-12T09:58:08.7449969Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-12T09:58:08.7469901Z ##[command]git config gc.auto 0
2019-11-12T09:58:08.7472048Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-12T09:58:08.7473824Z ##[command]git config --get-all http.proxy
2019-11-12T09:58:08.7477600Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66282/merge:refs/remotes/pull/66282/merge
---
2019-11-12T10:58:44.4467909Z .................................................................................................... 1400/9231
2019-11-12T10:58:51.0180404Z .................................................................................................... 1500/9231
2019-11-12T10:58:57.3861025Z .................................................................................................... 1600/9231
2019-11-12T10:59:07.0690486Z .................................................................................................... 1700/9231
2019-11-12T10:59:15.9130887Z ..i................................................................................................. 1800/9231
2019-11-12T10:59:22.9995859Z ......................................................................................iiiii......... 1900/9231
2019-11-12T10:59:45.3831654Z .................................................................................................... 2100/9231
2019-11-12T10:59:47.9291607Z .................................................................................................... 2200/9231
2019-11-12T10:59:50.5334209Z .................................................................................................... 2300/9231
2019-11-12T11:00:00.5237723Z .................................................................................................... 2400/9231
---
2019-11-12T11:03:01.2890719Z ..................................................................................i...............i. 4700/9231
2019-11-12T11:03:08.6602816Z .................................................................................................... 4800/9231
2019-11-12T11:03:18.3228370Z .................................................................................................... 4900/9231
2019-11-12T11:03:23.8782378Z .................................................................................................... 5000/9231
2019-11-12T11:03:35.9462867Z .....................................................................................ii.ii.......... 5100/9231
2019-11-12T11:03:39.9199628Z .i.................................................................................................. 5200/9231
2019-11-12T11:03:55.1276593Z .................................................................................................... 5400/9231
2019-11-12T11:04:02.5236450Z ...................................................................i................................ 5500/9231
2019-11-12T11:04:10.2134422Z .................................................................................................... 5600/9231
2019-11-12T11:04:18.4989027Z .................................................................................................... 5700/9231
2019-11-12T11:04:18.4989027Z .................................................................................................... 5700/9231
2019-11-12T11:04:27.9772630Z ....................................................ii...i..ii...........i.......................... 5800/9231
2019-11-12T11:04:51.7340015Z .................................................................................................... 6000/9231
2019-11-12T11:05:00.3577000Z .................................................................................................... 6100/9231
2019-11-12T11:05:00.3577000Z .................................................................................................... 6100/9231
2019-11-12T11:05:08.2372074Z .......................................................................i..ii........................ 6200/9231
2019-11-12T11:05:38.6274215Z .................................................................................................... 6400/9231
2019-11-12T11:05:40.9449537Z .......................................i............................................................ 6500/9231
2019-11-12T11:05:43.2794877Z .................................................................................................... 6600/9231
2019-11-12T11:05:45.7077369Z .......................i............................................................................ 6700/9231
---
2019-11-12T11:11:15.6860171Z  finished in 6.075
2019-11-12T11:11:15.7061370Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-12T11:11:15.9079521Z 
2019-11-12T11:11:15.9088724Z running 156 tests
2019-11-12T11:11:18.9331223Z iiii....iii......iii..iiii...i.............................i..i..................iF...i...........ii 100/156
2019-11-12T11:11:20.9184634Z .i.i..iiii..............i.........iii.i.........ii......
2019-11-12T11:11:21.4772220Z 
2019-11-12T11:11:21.4773095Z ---- [codegen] codegen/match.rs stdout ----
2019-11-12T11:11:21.4773293Z 
2019-11-12T11:11:21.4773293Z 
2019-11-12T11:11:21.4773800Z error: verification with 'FileCheck' failed
2019-11-12T11:11:21.4773979Z status: exit code: 1
2019-11-12T11:11:21.4774600Z command: "/usr/lib/llvm-6.0/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll" "/checkout/src/test/codegen/match.rs"
2019-11-12T11:11:21.4776023Z ------------------------------------------
2019-11-12T11:11:21.4776203Z 
2019-11-12T11:11:21.4776543Z ------------------------------------------
2019-11-12T11:11:21.4777032Z stderr:
2019-11-12T11:11:21.4777032Z stderr:
2019-11-12T11:11:21.4778005Z ------------------------------------------
2019-11-12T11:11:21.4778233Z /checkout/src/test/codegen/match.rs:13:11: error: expected string not found in input
2019-11-12T11:11:21.4778832Z // CHECK: switch{{.*}}, label %[[OTHERWISE:[a-zA-Z0-9_]+]] [
2019-11-12T11:11:21.4779119Z           ^
2019-11-12T11:11:21.4779621Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll:7:30: note: scanning from here
2019-11-12T11:11:21.4779814Z define void @exhaustive_match(i1 zeroext) unnamed_addr #0 {
2019-11-12T11:11:21.4779977Z                              ^
2019-11-12T11:11:21.4780475Z /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/match/match.ll:11:13: note: possible intended match here
2019-11-12T11:11:21.4780867Z  store i8 %1, i8* %e, align 1
2019-11-12T11:11:21.4781473Z 
2019-11-12T11:11:21.4781770Z ------------------------------------------
2019-11-12T11:11:21.4781901Z 
2019-11-12T11:11:21.4782004Z 
---
2019-11-12T11:11:21.4783430Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-12T11:11:21.4783599Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-12T11:11:21.4783704Z 
2019-11-12T11:11:21.4783801Z 
2019-11-12T11:11:21.4786604Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/codegen" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "codegen" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-12T11:11:21.4787633Z 
2019-11-12T11:11:21.4787757Z 
2019-11-12T11:11:21.4787908Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-12T11:11:21.4788080Z Build completed unsuccessfully in 1:06:50
2019-11-12T11:11:21.4788080Z Build completed unsuccessfully in 1:06:50
2019-11-12T11:11:21.4788222Z == clock drift check ==
2019-11-12T11:11:21.4788366Z   local time: Tue Nov 12 11:11:20 UTC 2019
2019-11-12T11:11:21.4788531Z   network time: Tue, 12 Nov 2019 11:11:21 GMT
2019-11-12T11:11:21.4788670Z == end clock drift check ==
2019-11-12T11:11:25.1920975Z 
2019-11-12T11:11:25.2033657Z ##[error]Bash exited with code '1'.
2019-11-12T11:11:25.2113471Z ##[section]Starting: Checkout
2019-11-12T11:11:25.2115252Z ==============================================================================
2019-11-12T11:11:25.2115309Z Task         : Get sources
2019-11-12T11:11:25.2115356Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented:
Adjusted the existing `match.rs` codegen test and introduced a new one which demonstrates that the MIR optimizations introduced produce better LLVM IR as compared to [before](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e703da8ad0506a99ed19dc566081869f).

---------------------

Me and Oliver have been discussing some generalizations to these optimizations (for follow ups). In particular, we can generalize the second one into two passes:

1. `DedupBlocks` -- This one looks for identical basic blocks, collects a substitution map, and then substitutes all references to `BasicBlock` for the key in the map. If we want, this could be run to a fixed point, but we could start it off as a one-shot pass.

2. `SwitchToSamePlace` -- This does (could be split up but maybe not depending on complexity):
    a. Filters BBs that are unreachable
    b. Merges identical targets
    c. If one target is left ==> goto
    d. If no targets are left ==> unreachable

   -- a. & d. sorta want to be run to a fixed point but we should probably start off as one-shot instead.

Finally, the `SimplifyArmIdentity` could be generalized to also recognize e.g. `Var(a, b, c) => Var(a, b, c)` and not just `Var(a) => Var(a)`.

Centril commented:
Actually, looking at `CfgSimplifier::simplify_branch` it looks like `SwitchToSamePlace` is a generalization of that logic. `simplify_branch` requires *all* targets to end up in the same place whereas the generalization would first do a filter and simplify e.g. `-> [bb2, bb3, otherwise: bb2]` to `-> [bb2, otherwise: bb3]`.

Centril commented:
This should be ready now :)

Centril commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,192 @@
+            let mut iter_bbs_reachable = targets
+                .iter()
+                .map(|idx| (*idx, &bbs[*idx]))
+                .filter(|(_, bb)| bb.terminator().kind != TerminatorKind::Unreachable)
```
@rkruppe So we would need to check that there are no `InlineAsm` statements here wrt. the point you made on the other PR?

rkruppe replied:
Yes. Though since there's at least two passes that need this information, I would recommend again to implementing that in a way that can be shared between the passes.

BTW, I'm starting to think it might be more elegant to treat this as a canonicalization rather than a check. That is, instead of repeatedly checking whether there's any inline asm, have a rewrite that removes non-divering statements leading up to a `Unreachable`. Then this and other passes can run the canonicalization (which can be made very cheap if nothing changed) and then just check for `bb.statements().is_empty() && bb.terminator().kind == Unreachable`.

oli-obk commented on src/test/codegen/try_identity.rs:
```diff
@@ -0,0 +1,17 @@
+// compile-flags: -C no-prepopulate-passes -Z mir-opt-level=2
```
Does this also work without -Z flags? Is this directly exposed to users via `--release`? Or should that be done in a separate PR?

Centril replied:
No idea. =) `-Z mir-opt-level=2` needed to trigger inlining, which this test needs. cc @wesleywiser 

We could maybe special case inlining for some functions via attributes?

wesleywiser replied:
I'm not sure the inline pass is stable enough to offer even via an opt-in attribute.

Centril replied:
@wesleywiser just to clarify, I meant `#[rustc_inline_me]` just for use in the standard library or something.

wesleywiser replied:
@Centril Yeah, that's what I thought you meant. I just think the inliner has a lot of sharp edges and I'd rather not give people tools *this* sharp.

Centril replied:
@wesleywiser Fair enough; Is there a dumber "definitely correct" version of the inliner we could use for "just this function" which doesn't try to inline recursively and whatnot?

wesleywiser replied:
Possibly yeah :)

bors commented:
:umbrella: The latest upstream changes (presumably #66384) made this pull request unmergeable. Please resolve the merge conflicts.

Centril commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,201 @@
+            let mut iter_bbs_reachable = targets
+                .iter()
+                .map(|idx| (*idx, &bbs[*idx]))
+                .filter(|(_, bb)| {
```
@oli-obk I adjusted this check to preserve soundness as discussed. I think the PR should be ready to go now. It's still useful even without `-Z mir-opt-level 2` for `match`es written out and such.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14043&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14043/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14043/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-21T00:57:59.3261254Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-21T00:58:00.3324008Z ##[command]git config gc.auto 0
2019-11-21T00:58:00.3328126Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-21T00:58:00.3331353Z ##[command]git config --get-all http.proxy
2019-11-21T00:58:00.3335422Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66282/merge:refs/remotes/pull/66282/merge
---
2019-11-21T01:58:36.9237642Z .................................................................................................... 1500/9260
2019-11-21T01:58:42.9181437Z .................................................................................................... 1600/9260
2019-11-21T01:58:51.3027927Z .................................................................................................... 1700/9260
2019-11-21T01:59:00.5952139Z .........i.......................................................................................... 1800/9260
2019-11-21T01:59:07.1955067Z ..............................................................................................iiiii. 1900/9260
2019-11-21T01:59:28.6634871Z .................................................................................................... 2100/9260
2019-11-21T01:59:31.1325232Z .................................................................................................... 2200/9260
2019-11-21T01:59:33.6574528Z .................................................................................................... 2300/9260
2019-11-21T01:59:39.9168631Z .................................................................................................... 2400/9260
---
2019-11-21T02:02:47.6017868Z ...........i........................................................................................ 4800/9260
2019-11-21T02:02:58.4453173Z .................................................................................................... 4900/9260
2019-11-21T02:03:04.0105843Z .................................................................................................... 5000/9260
2019-11-21T02:03:14.9640937Z .................................................................................................... 5100/9260
2019-11-21T02:03:20.5845974Z .ii.ii...........i.................................................................................. 5200/9260
2019-11-21T02:03:31.2678923Z .................................................................................................... 5400/9260
2019-11-21T02:03:41.9116326Z ...................................................................................i................ 5500/9260
2019-11-21T02:03:50.0899477Z .................................................................................................... 5600/9260
2019-11-21T02:03:56.3824322Z .................................................................................................... 5700/9260
2019-11-21T02:03:56.3824322Z .................................................................................................... 5700/9260
2019-11-21T02:04:06.8126331Z .....................................................................ii...i..ii...........i......... 5800/9260
2019-11-21T02:04:29.9599642Z .................................................................................................... 6000/9260
2019-11-21T02:04:38.5231221Z .................................................................................................... 6100/9260
2019-11-21T02:04:38.5231221Z .................................................................................................... 6100/9260
2019-11-21T02:04:43.4094944Z ..........................................................................................i..ii..... 6200/9260
2019-11-21T02:05:07.7162202Z .................................................................................................... 6400/9260
2019-11-21T02:05:16.3577697Z ..........................................................i......................................... 6500/9260
2019-11-21T02:05:18.7214854Z .................................................................................................... 6600/9260
2019-11-21T02:05:21.1028671Z ...............................................i.................................................... 6700/9260
---
2019-11-21T02:10:22.4549959Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-21T02:10:22.4550465Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-21T02:10:22.4567336Z 
2019-11-21T02:10:22.4567625Z 
2019-11-21T02:10:22.4569665Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-21T02:10:22.4570427Z 
2019-11-21T02:10:22.4570474Z 
2019-11-21T02:10:22.4577513Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-21T02:10:22.4577603Z Build completed unsuccessfully in 1:06:01
2019-11-21T02:10:22.4577603Z Build completed unsuccessfully in 1:06:01
2019-11-21T02:10:22.4635414Z == clock drift check ==
2019-11-21T02:10:22.4651152Z   local time: Thu Nov 21 02:10:22 UTC 2019
2019-11-21T02:10:22.7463633Z   network time: Thu, 21 Nov 2019 02:10:22 GMT
2019-11-21T02:10:22.7464642Z == end clock drift check ==
2019-11-21T02:10:23.6619398Z 
2019-11-21T02:10:23.6733561Z ##[error]Bash exited with code '1'.
2019-11-21T02:10:23.6781974Z ##[section]Starting: Checkout
2019-11-21T02:10:23.6783477Z ==============================================================================
2019-11-21T02:10:23.6783522Z Task         : Get sources
2019-11-21T02:10:23.6783561Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented:
Messed up the condition in ^--. All fixed now.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit eeba2ec9760e90a1ba999be95e488f12b07766c2 has been approved by `oli-obk`

<!-- @bors r=oli-obk eeba2ec9760e90a1ba999be95e488f12b07766c2 -->
<!-- homu: {"type":"Approved","sha":"eeba2ec9760e90a1ba999be95e488f12b07766c2","approver":"oli-obk"} -->

Centril commented:
@bors rollup=never (soundness...)

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout simplify-try` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self simplify-try --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/librustc_mir/lib.rs
CONFLICT (content): Merge conflict in src/librustc_mir/lib.rs
Auto-merging src/librustc/mir/mod.rs
Auto-merging src/librustc/hir/mod.rs
CONFLICT (content): Merge conflict in src/librustc/hir/mod.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #66607) made this pull request unmergeable. Please resolve the merge conflicts.

Centril commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 2f00e86cb5cbb4d4cfe17abc6136aacadbe02382 has been approved by `oli-obk`

<!-- @bors r=oli-obk 2f00e86cb5cbb4d4cfe17abc6136aacadbe02382 -->
<!-- homu: {"type":"Approved","sha":"2f00e86cb5cbb4d4cfe17abc6136aacadbe02382","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 2f00e86cb5cbb4d4cfe17abc6136aacadbe02382 with merge ae731e69b0e519dbf73d4a870a82224b778808de...
<!-- homu: {"type":"BuildStarted","head_sha":"2f00e86cb5cbb4d4cfe17abc6136aacadbe02382","merge_sha":"ae731e69b0e519dbf73d4a870a82224b778808de"} -->

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14112&view=logs&jobId=34a2f069-38a6-567f-c087-79a147504c7a) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14112/logs/1159), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14112/logs/1159)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-22T00:10:43.0377317Z do so (now or later) by using -b with the checkout command again. Example:
2019-11-22T00:10:43.0378736Z 
2019-11-22T00:10:43.0378832Z   git checkout -b <new-branch-name>
2019-11-22T00:10:43.0378867Z 
2019-11-22T00:10:43.0378942Z HEAD is now at ae731e69b Auto merge of #66282 - Centril:simplify-try, r=oli-obk
2019-11-22T00:10:43.0769463Z ##[section]Starting: Decide whether to run this job
2019-11-22T00:10:43.0870334Z ==============================================================================
2019-11-22T00:10:43.0870455Z Task         : Bash
2019-11-22T00:10:43.0870535Z Description  : Run a Bash script on macOS, Linux, or Windows
---
2019-11-22T00:10:44.1732643Z 
2019-11-22T00:10:44.1732668Z 
2019-11-22T00:10:44.1732692Z 
2019-11-22T00:10:44.1732733Z 
2019-11-22T00:10:44.1732799Z     Err(x) => Err(x),
2019-11-22T00:10:44.1732868Z     Ok(x) => Ok(x),
2019-11-22T00:10:44.1732929Z   ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
2019-11-22T00:10:44.1733006Z   _LOCAL_0 = move _LOCAL_1
2019-11-22T00:10:44.1733069Z   _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
2019-11-22T00:10:44.1733140Z   ```
2019-11-22T00:10:44.1733182Z   ```
2019-11-22T00:10:44.1733241Z   ```rust
2019-11-22T00:10:44.1733326Z   ```rust
2019-11-22T00:10:44.1733373Z   discriminant(_LOCAL_0) = VAR_IDX;
2019-11-22T00:10:44.1733598Z   into:
2019-11-22T00:10:44.1733653Z - `SimplifyArmIdentity` which transforms something like:
2019-11-22T00:10:44.1733752Z - `SimplifyBranchSame` which transforms `SwitchInt`s to identical basic blocks into a `goto` to the first reachable target.
2019-11-22T00:10:44.1733902Z AGENT_DISABLELOGPLUGIN_TESTFILEPUBLISHERPLUGIN=true
2019-11-22T00:10:44.1733982Z AGENT_DISABLELOGPLUGIN_TESTRESULTLOGPLUGIN=true
2019-11-22T00:10:44.1734038Z AGENT_HOMEDIRECTORY=C:\agents\2.160.1
2019-11-22T00:10:44.1734106Z AGENT_ID=520
---
2019-11-22T00:10:44.1741938Z BUILD_SOURCEBRANCHNAME=auto
2019-11-22T00:10:44.1742005Z BUILD_SOURCESDIRECTORY=D:\a\1\s
2019-11-22T00:10:44.1742068Z BUILD_SOURCEVERSION=ae731e69b0e519dbf73d4a870a82224b778808de
2019-11-22T00:10:44.1742142Z BUILD_SOURCEVERSIONAUTHOR=bors
2019-11-22T00:10:44.1742230Z BUILD_SOURCEVERSIONMESSAGE=Auto merge of #66282 - Centril:simplify-try, r=oli-obk
2019-11-22T00:10:44.1742473Z CI_JOB_NAME=x86_64-msvc-cargo
2019-11-22T00:10:44.1742526Z COBERTURA_HOME=C:\cobertura-2.1.1
2019-11-22T00:10:44.1742657Z COMMONPROGRAMFILES=C:\Program Files\Common Files
2019-11-22T00:10:44.1742722Z COMMON_TESTRESULTSDIRECTORY=D:\a\1\TestResults
---
2019-11-22T00:10:44.1746232Z HOMEPATH=\Users\VssAdministrator
2019-11-22T00:10:44.1746312Z IEWebDriver=C:\SeleniumWebDrivers\IEDriver
2019-11-22T00:10:44.1746369Z INPUT_ARGUMENTS=
2019-11-22T00:10:44.1748921Z ImageVersion=20191028.1
2019-11-22T00:10:44.1749288Z It should be noted however that the desugaring of `?` includes a function call and so the first pass in this PR relies on inlining to substitute that function call for identity on `x`. Inlining requires `mir-opt-level=2` so this might not have any effect in perf-bot but let's find out.
2019-11-22T00:10:44.1749792Z JAVA_HOME_11_X64=C:\Program Files\Java\zulu-11-azure-jdk_11.33.15-11.0.4-win_x64
2019-11-22T00:10:44.1749876Z JAVA_HOME_7_X64=C:\Program Files\Java\zulu-7-azure-jdk_7.31.0.5-7.0.232-win_x64
2019-11-22T00:10:44.1749972Z JAVA_HOME_8_X64=C:\Program Files\Java\zulu-8-azure-jdk_8.40.0.25-8.0.222-win_x64
2019-11-22T00:10:44.1750060Z LOCALAPPDATA=C:\Users\VssAdministrator\AppData\Local
---
2019-11-22T00:10:44.1790457Z TMP=/tmp
2019-11-22T00:10:44.1790526Z TOOLSTATE_ISSUES_API_URL=https://api.github.com/repos/rust-lang/rust/issues
2019-11-22T00:10:44.1790601Z TOOLSTATE_PUBLISH=1
2019-11-22T00:10:44.1790664Z TOOLSTATE_REPO=https://github.com/rust-lang-nursery/rust-toolstate
2019-11-22T00:10:44.1790759Z This PR works towards https://github.com/rust-lang/rust/issues/66234 by providing two optimization passes meant to run in sequence:
2019-11-22T00:10:44.1790845Z Together, these are meant to simplify the following into just `res`:
2019-11-22T00:10:44.1791153Z USERDOMAIN=fv-az665
2019-11-22T00:10:44.1791225Z USERDOMAIN_ROAMINGPROFILE=fv-az665
2019-11-22T00:10:44.1791297Z USERNAME=VssAdministrator
2019-11-22T00:10:44.1791353Z USERPROFILE=C:\Users\VssAdministrator
2019-11-22T00:10:44.1791353Z USERPROFILE=C:\Users\VssAdministrator
2019-11-22T00:10:44.1791431Z VCPKG_INSTALLATION_ROOT=C:\vcpkg
2019-11-22T00:10:44.1791487Z VCVARS_BAT=vcvars64.bat
2019-11-22T00:10:44.1791566Z VS140COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\
2019-11-22T00:10:44.1791634Z VSTS_AGENT_PERFLOG=c:\vsts\perflog
2019-11-22T00:10:44.1791717Z VSTS_PROCESS_LOOKUP_ID=vsts_2d5c59f1-62fe-4ead-a819-c4591a3bcde2
2019-11-22T00:10:44.1791794Z WINDIR=C:\windows
2019-11-22T00:10:44.1791852Z WIX=C:\Program Files (x86)\WiX Toolset v3.11\
2019-11-22T00:10:44.1791934Z [mir-opt] asking `?`s in a more optimized fashion
2019-11-22T00:10:44.1792059Z ```
2019-11-22T00:10:44.1792104Z ```rust
2019-11-22T00:10:44.1792166Z agent.jobstatus=Succeeded
2019-11-22T00:10:44.1792219Z match res {
2019-11-22T00:10:44.1792219Z match res {
2019-11-22T00:10:44.1792301Z r? @oli-obk -- This is WIP, but I'd appreciate feedback. :)
2019-11-22T00:10:44.1792438Z 
2019-11-22T00:10:44.1796114Z disk usage:
2019-11-22T00:10:44.2281915Z Filesystem            Size  Used Avail Use% Mounted on
2019-11-22T00:10:44.2282059Z C:/Program Files/Git  256G  140G  116G  55% /
---
2019-11-22T00:10:55.3411300Z  30  480M   30  146M    0     0  22.1M      0  0:00:21  0:00:06  0:00:15 22.3M
2019-11-22T00:10:56.7898963Z  36  480M   36  173M    0     0  22.7M      0  0:00:21  0:00:07  0:00:14 23.3M
2019-11-22T00:10:56.8221928Z  38  480M   38  186M    0     0  20.6M      0  0:00:23  0:00:09  0:00:14 20.1M
2019-11-22T00:10:56.8249797Z  39  480M   39  187M    0     0  20.6M      0  0:00:23  0:00:09  0:00:14 20.0M
2019-11-22T00:10:56.8250879Z curl: (56) OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 10054
2019-11-22T00:10:56.8271581Z 
2019-11-22T00:10:56.8271869Z gzip: stdin: unexpected end of file
2019-11-22T00:10:56.8277445Z tar: Unexpected EOF in archive
2019-11-22T00:10:56.8277542Z tar: Unexpected EOF in archive
2019-11-22T00:10:56.8277621Z tar: Error is not recoverable: exiting now
2019-11-22T00:10:56.8334979Z 
2019-11-22T00:10:56.8412459Z ##[error]Bash exited with code '2'.
2019-11-22T00:10:56.8597332Z ##[section]Starting: Checkout
2019-11-22T00:10:56.8693690Z ==============================================================================
2019-11-22T00:10:56.8693798Z Task         : Get sources
2019-11-22T00:10:56.8693864Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14112)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14112","builder_name":"checks-azure"} -->

Centril commented:
@bors retry spurious

bors commented:
:hourglass: Testing commit 2f00e86cb5cbb4d4cfe17abc6136aacadbe02382 with merge abd69551bf8b8755b5e00d4f4d45ae5d4a0cd17d...
<!-- homu: {"type":"BuildStarted","head_sha":"2f00e86cb5cbb4d4cfe17abc6136aacadbe02382","merge_sha":"abd69551bf8b8755b5e00d4f4d45ae5d4a0cd17d"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14114)
Approved by: oli-obk
Pushing abd69551bf8b8755b5e00d4f4d45ae5d4a0cd17d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14114"},"merge_sha":"abd69551bf8b8755b5e00d4f4d45ae5d4a0cd17d"} -->

Dylan-DPC commented:
@oli-obk the linked pr has been merged already. Unblocking this

Dylan-DPC commented:
@oli-obk sorry wrong pr 😂

eddyb commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,201 @@
+                || Some((local_0, vf_s0.var_idx)) != match_set_discr(s2)
+            {
+                continue;
+            }
```
You're not checking that `local_0 != local_1`, which is another soundness condition.
cc @rust-lang/wg-mir-opt

eddyb commented on src/librustc_mir/transform/simplify_try.rs:
````diff
@@ -0,0 +1,201 @@
+///
+/// ```rust
+/// _LOCAL_TMP = ((_LOCAL_1 as Variant ).FIELD: TY );
+/// ((_LOCAL_0 as Variant).FIELD: TY) = move _LOCAL_TMP;
````
This is only sound to collapse if `_LOCAL_TMP` isn't used *anywhere* else (see `promote_consts` for an example of an analysis like that).

But also, there's another issue here... namely, debuginfo.

You need to replace `_LOCAL_TMP` with `((_LOCAL_1 as Variant ).FIELD: TY)` in debuginfo as well, making this a very specialized form of copy-prop, I guess?

cc @rust-lang/wg-mir-opt

eddyb replied:
It would be nice to preserve debuginfo perfectly, but for that we might need a way to merge scopes or something, so that `SourceInfo` from all 3 statements is combined.

Effectively, the resulting `_LOCAL_0 = move _LOCAL_1` needs to be in a superposition of all the scopes.

Which might be hard due to the inlining.

eddyb commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,201 @@
+                _ => unreachable!(),
+            }
+            s1.make_nop();
+            s2.make_nop();
```
I would replace the third statement and nop the first two. That way, you get the `SourceInfo` of the `Ok(x)` rather than the `x` use (or whatever initialized it).

eddyb commented on src/librustc_mir/transform/simplify_try.rs:
```diff
@@ -0,0 +1,201 @@
+/// Simplifies `SwitchInt(_) -> [targets]`,
+/// where all the `targets` have the same form,
+/// into `goto -> target_first`.
+pub struct SimplifyBranchSame;
```
I don't think we can do this unless there is no debuginfo attached to scopes used only in the arms, *or* we can convert the debuginfo into a form that switches on the `SwitchInt` operand (plausible in DWARF, not really exposed in LLVM).

eddyb replied:
But also this optimization can be rephrased as sinking identical statements from predecessors into the start of a block, followed by `simplify_cfg`.
